tests/lib/tpp_schema.py

```diff
@@ -1,1057 +1,3261 @@
+    Carer_Support_Indicator = mapped_column(t.VARCHAR(1000))
+    Der_Activity_Month = mapped_column(t.VARCHAR(6))
+    Der_Admit_Treatment_Function_Code = mapped_column(t.VARCHAR(3))
+    Der_Diagnosis_All = mapped_column(t.VARCHAR(4000))
+    Der_Diagnosis_Count = mapped_column(t.Integer)
+    Der_Dischg_Treatment_Function_Code = mapped_column(t.VARCHAR(3))
+    Der_Financial_Year = mapped_column(t.VARCHAR(7))
+    Der_Procedure_All = mapped_column(t.VARCHAR(4000))
+    Der_Procedure_Count = mapped_column(t.Integer)
+    Der_Pseudo_Patient_Pathway_ID = mapped_column(t.BIGINT)
+    Der_Spell_LoS = mapped_column(t.Integer)
+    Discharge_Date = mapped_column(t.Date)
+    Discharge_Destination = mapped_column(t.VARCHAR(2))
+    Discharge_Method = mapped_column(t.VARCHAR(2))
+    Duration_of_Elective_Wait = mapped_column(t.Integer)
+    Ethnic_Group = mapped_column(t.VARCHAR(2))
+    Hospital_Spell_Duration = mapped_column(t.Integer)
+    Patient_Classification = mapped_column(t.VARCHAR(2))
+    Provider_Org_Code_Type = mapped_column(t.VARCHAR(5))
+    Source_of_Admission = mapped_column(t.VARCHAR(2))
+    Spell_Core_HRG_SUS = mapped_column(t.VARCHAR(10))
+    Spell_HRG_Version_No_SUS = mapped_column(t.VARCHAR(20))
+
+
+class APCS_Cost(Base):
+    __tablename__ = "APCS_Cost"
+    _pk = mapped_column(t.Integer, primary_key=True)
+
+    Patient_ID = mapped_column(t.BIGINT)
+    APCS_Ident = mapped_column(t.BIGINT)
+    Grand_Total_Payment_MFF = mapped_column(t.REAL)
+    Tariff_Initial_Amount = mapped_column(t.REAL)
+    Tariff_Total_Payment = mapped_column(t.REAL)
 
 
-class MedicationIssue(Base):
-    __tablename__ = "MedicationIssue"
+class APCS_Der(Base):
+    __tablename__ = "APCS_Der"
+    _pk = mapped_column(t.Integer, primary_key=True)
+
+    Patient_ID = mapped_column(t.BIGINT)
+    APCS_Ident = mapped_column(t.BIGINT)
+    Spell_Dominant_Procedure = mapped_column(t.VARCHAR(100))
+    Spell_LoS = mapped_column(t.VARCHAR(5))
+    Spell_Main_Specialty_Code = mapped_column(t.VARCHAR(3))
+    Spell_PbR_CC_Day = mapped_column(t.VARCHAR(4))
+    Spell_PbR_Rehab_Days = mapped_column(t.VARCHAR(4))
+    Spell_Primary_Diagnosis = mapped_column(t.VARCHAR(5))
+    Spell_RE30_Admit_Type = mapped_column(t.VARCHAR(1))
+    Spell_RE30_Indicator = mapped_column(t.VARCHAR(1))
+    Spell_Secondary_Diagnosis = mapped_column(t.VARCHAR(5))
+    Spell_Treatment_Function_Code = mapped_column(t.VARCHAR(3))
 
-    Patient_ID = Column(Integer, ForeignKey("Patient.Patient_ID"))
-    Patient = relationship("Patient", back_populates="MedicationIssues")
-    Consultation_ID = Column(Integer)
-    MedicationIssue_ID = Column(Integer, primary_key=True)
-    RepeatMedication_ID = Column(Integer)
-    MultilexDrug_ID = Column(
-        NVARCHAR(length=20), ForeignKey("MedicationDictionary.MultilexDrug_ID")
-    )
-    MedicationDictionary = relationship(
-        "MedicationDictionary", back_populates="MedicationIssues", cascade="all, delete"
-    )
```

Comment:
question: Why can we remove the use of some of the explicitly specified primary and foreign keys, and relationships? Is this just because SQLAlchemy 2 lets us do this?

I'm sure this is fine; just for my own understanding :smiley:

Reply:
This is discussed a bit in the original issue but I didn't reference this at in the PR:

> Additionally, the ORM models in `tpp_schema.py` are much more complex than they need to be as they define a whole load of foreign key relationships between the different models. This is because they were copied directly from Cohort Extractor where they formed the primary means of constructing test fixtures. It was therefore important to be able easily to construct complex sets of interrelated objects. This need isn't so great in Data Builder as the models exist purely to test the `QueryTable` and `MappedTable` definitions.

So the answer is that the foreign key relationships were things we added for our own convenience, not things represented in the database itself.

Reply:
Thanks :+1: